Documentation/design-docs/COM-activation.md

```diff
@@ -0,0 +1,138 @@
+1) The `IClassFactory` instance will be returned to the caller of `DllGetClassObject()`.
+
+The `DllCanUnloadNow()` function will always return `S_FALSE` indicating the shim is never able to be unloaded. This matches .NET Framework semantics and can be adjusted in the future if needed.
+
+#### Class Registration
+
+Registration will depend on the application's [deployment scenario](https://docs.microsoft.com/en-us/dotnet/core/deploying/).
+
+* [Framework-dependent deployments (FDD)](https://docs.microsoft.com/en-us/dotnet/core/deploying/#framework-dependent-deployments-fdd) - This is most similar to the .NET Framework scenario and registration will be done using the shared framework. 
+* [Self-contained deployments (SCD)](https://docs.microsoft.com/en-us/dotnet/core/deploying/#self-contained-deployments-scd) - Registration will be done using the contained framework.
+
+##### Registry
+
+Details of class registration in the registry for .NET Core are under active investigation, but this fact does not limit the current plan. At a minimum, registration scripts could be generated for the user via an extension to the [`dotnet.exe`](https://docs.microsoft.com/en-us/dotnet/core/tools/dotnet?tabs=netcore21) tool, although this approach may be unsatisfactory as a final experience.
+
+##### Registration-Free
+
+[RegFree COM for .NET](https://docs.microsoft.com/en-us/dotnet/framework/interop/configure-net-framework-based-com-components-for-reg) is another style of registration that does not require accessing the registry. This approach is complicated by the use of [application manifests](https://docs.microsoft.com/en-us/windows/desktop/SbsCs/application-manifests), but does have benefits for limiting environment impact and simplifying deployment. A severe limitation of this approach is that in order to use RegFree COM with a .NET class, the Window OS assumes the use of `mscoree.dll` for the in-proc server. Without a change in the Windows OS, this assumption in the RegFree .NET scenario makes usage of the current manifest approach a broken scenario for .NET Core.
+
+An example of a RegFree manifest for a .NET Framework class is below - note the absence of specifying a hosting server library (i.e. `mscoree.dll` is implied for the `clrClass` element).
+
+``` xml
+<?xml version="1.0" encoding="UTF-8" standalone="yes"?>
+<assembly xmlns="urn:schemas-microsoft-com:asm.v1" manifestVersion="1.0">
+    <assemblyIdentity
+        type="win32"
+        name="NetServer"
+        version="1.0.0.0">
+    </assemblyIdentity>
+    <clrClass
+        clsid="{3C58BBC9-3966-4B58-8EE2-398CBBC9FDC4}"
+        name="NetServer.Server"
+        threadingModel="Both"
+        runtimeVersion="v4.0.30319">
+    </clrClass>
+</assembly>
+```
+
+Due to the above issues with traditional RegFree manifests and .NET classes, an alternative system must be employed to enable a low-impact style of class registration for .NET Core.
```

Comment:
Wouldn't a bog-standard `<comClass />` SxS manifest entry work here? It wouldn't be able to specify the CLR class information, but it would allow the specification of the server DLL. Furthermore, since COM is by definition Windows only, would it be possible to embed the CLSID-to-CLR-class information as a native resource in the DLL that is listed in the manifest? This is very easy to do using .NET; I use the [ResourceLib](https://github.com/resourcelib/resourcelib) library to do this usually, but you could P/Invoke `UpdateResource()` and friends manually from kernel32 if you really wanted.

I don't think that doing this would affect SCD/FDD any, as the DLL that is listed in the manifest must be present in the same directory hierarchy as the manifest (because Windows is too stupid to look anywhere else). The DLL that is listed in the manifest can then resolve the managed assemblies any way it sees fit.

Reply:
> Wouldn't a bog-standard <comClass /> SxS manifest entry work here?

Potentially, but it is a bit more complicated than that. The user shouldn't be expected to author their own server, so it would need to be provided - which is the shim. However, the shim's manifest would then need to declare the `GUID` of the class it intends to provide so SxS would discover the server. Basically this approach is what I am doing and the broad strokes of your suggestion are valid, but like everything in COM the devil is in the details. I will updating this doc in the next few days.

> because Windows is too stupid to look anywhere else
That is actually because of a very complex security scenario and was intentionally designed that way.

Reply:
@wjk This is an example of what it sounds like you are suggesting: https://github.com/AaronRobinsonMSFT/COMInterop/tree/coreclr_example. It is what I am currently investigating as a possible solution. Let me know what you think. Thanks.

Reply:
Looks perfect! All we would really need to do then is to hook this code up to the `*.deps.json` mechanism (because I don't think it can parse that as-is). Furthermore, with a bit of expansion, this approach could well implement the out-of-proc COM hosting I was asking about in https://github.com/dotnet/coreclr/issues/18490. Cool beans!

Reply:
>All we would really need to do then is to hook this code up to the *.deps.json mechanism (because I don't think it can parse that as-is).

I am unclear at the moment if we would use `runtimeconfig.json` or `deps.json`. There has been much talk about that. @vitek-karas should be able to provide some better insight into those arguments though. The current implementation is basically how I am planning on adding support in the coreclr repo for adding .NET activation testing, as you can imagine actual product support is much more complicated. 

>... this approach could well implement the out-of-proc COM hosting ...

Out of proc COM hosting is definitely up in the air with respect to support in .NET Core. There are some real complex scenarios regarding proxies and I don't know if/how they would fit into the overall .NET Core platform itself. Definitely doable, but would require a significant investment to add support for out-of-proc even if we have in-proc.